### ISSUE TITLE
Not possible to mock relative paths after upgrading to 0.4.x

### ISSUE THREAD

**Issue by birtles** [IanVS/vitest-fetch-mock#31]
I might not have properly analyzed this, but version 0.4.1 is not working in a couple of projects and I think it comes down to handling of relative URLs.

My projects run in a Web / Web extension context so there's a defined base URL to resolve relative URLs against. As a result `fetch('folder/file.json')` is valid and I've been mocking that in my tests.

However, it seems like with the switch from cross-fetch to native Node fetch, we force the `fetch` parameters into a native `Request` object here: https://github.com/IanVS/vitest-fetch-mock/blob/66aeefd26577e48693df2c79dd232cb8217fbebb/src/index.ts#L312

Apparently, Node's implementation of the `Request` constructor doesn't handle relative URLs since there's no base URL to resolve against and this will throw with a `TypeError` ("Failed to parse URL from folder/file.json"). As a result the fetch fails.

I'm not really sure what the correct solution is. I can probably work around it in my projects by detecting when we are running tests and supplying a dummy base URL but it would be nice if that wasn't necessary and we could maintain compatibility with 0.3.0. Perhaps we could wrap native `Request` somehow?

**Comment by dirkluijk** [comment]
This really is a bummer. It's a bit silly that constructing a `Request` with a relative URL is prohibited in Node.js. I would expect this error only to be thrown when the Fetch API is actually invoked.

I have several solutions in mind:

1) detect any non-absolute URL and convert them into absolute ULRs `vitest-fetch-mock`; e.g. `foo` becoming `http://127.0.0.1/foo`. However, this trick would not entirely be "invisible" to users, since the `Request` object is passed to the `ResponseProvider` function argument of `vitest-fetch-mock`. This seems to be solution with the least effort, but we need to investigate the impact.

2) use a custom `Request` class, e.g. from a polyfill like `cross-fetch`. This is rather annoying though, since this would introduce another dependency that we just got rid of. Note that other polyfills like `node-fetch` also seem to have the same limitation. Alternatively, we could extend `Request` ourselves:

```ts
/**
 * A custom Request implementation that permits relative URLs
 * to enable non-Node.js use cases.
 */
class MockedRequest extends Request {
    private inputUrl: string;

    constructor(input: RequestInput, init?: RequestInit) {
        const inputUrl = input instanceof Request ? input.url : input instanceof URL ? input.toString() : input;
        const patchedInput = input instanceof Request ? input : input instanceof URL ? toAbsolute(input.toString()) : toAbsolute(inputUrl);
        super(patchedInput, init);
        this.inputUrl = inputUrl;
    }

    // @ts-expect-error We know what we're doing
    override get url(): string {
        return this.inputUrl;
    }
}

/**
 * Converts non-absolute URLs into absolute URLs
 * @param url
 */
function toAbsolute(url: string) {
    if (!new RegExp('^(?:[a-z+]+:)?//', 'i').test(url)) {
        return `http://127.0.0.1/${url}`;
    }
    return url;
}
```

3) we change `vitest-fetch-mock` to stay away from constructing actual `Request` objects by changing the normalization step. This would however lead to breaking changes, like changing the `ResponseProvider` interface to accept a different "response like" object instead of the actual `Response`. At this point, I'm would not be in favor of this since it's a lot of effort with a lot of impact.

@IanVS what is your opinion?

**Comment by IanVS** [comment]
Bummer indeed.  I think we could probably take the same approach as MSW here: https://github.com/mswjs/interceptors/commit/f9193cd74602bbd78c30086651b6a456d8662fd0, using `location.origin` when possible.  This will mean that users will need to have something like jsdom in place, but if they're running browser tests in node, they should already.  (Actually they should use Vitest browser mode instead, but that's another issue).

**Comment by dirkluijk** [comment]
Well, what about the Vitest browser mode? Isn't that exactly meant to solve environment-specific issues like this?

**Comment by IanVS** [comment]
Yes, though it is still marked as "experimental" and perhaps not all projects can use it.  I think this tool should still work with jsdom + node, since that's a much more common approach still.

**Comment by IanVS** [comment]
I'll work on this to try to unblock folks.  It looks like the :+1: are piling up.

**Comment by IanVS** [comment]
The fix has been released in `0.4.2`.
